Download Disputes CSV from service

shendy-a8c opened the pull request:
Fixes https://github.com/Automattic/transact-platform-server/issues/7188

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

~~Note: this PR requires some changes from https://github.com/Automattic/woocommerce-payments/pull/10211, so base branch is set to PR 10211's branch.~~ #10211 is now merged, targeting `develop` as normal

Get rid off in-browser CSV download for Payments > Disputes. With this PR, even though the number of rows shown on the Disputes page is low (<= 25), the CSV to be downloaded will be generated in the server. The browser will then do some polling to see if the CSV file is ready to be downloaded in the browser. The CSV file will also be emailed.

#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

Test this PR with https://github.com/Automattic/transact-platform-server/pull/7200

* Open Payments > Disputes.
* Click the Download button to download the CSV.
* Notice the CSV is not downloaded in the browser immediately. You will see a notice that the CSV will be emailed. If you're quick enough, keep refreshing one of the admin pages of your local WCPay server so the async job that will generate the CSV file is executed. If the file is ready before the polling attempt finishes, you will see the file gets downloaded in your browser.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

Commit:
Initial commit - a polling and fallback implementation on the export job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
added changelog

Jinksi committed:
Add useReportExport hook to encapsulate polling logic for PR 10211 (#10212)

Commit:
Bug fixes & Changes for DownloadURLResponse

- The response is changed to an array
- If the file times out, don't show an error. The file may be emailed.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Deleted tests relating to the removed JS CSV download functionality.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Added checks for status in response.

Commit:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Should handle the error if downloadURL API sends 500 error.

Commit:
Merge branch 'update/9969-poll-async-job' of https://github.com/Automattic/woocommerce-payments into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Removed some unused variables in the test file.

shendy-a8c committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Add test for report export hook

Commit:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Merge branch 'develop' into update/9969-poll-async-job

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Remove labelInCsv as it was used only for the
JS download functionality.

Commit:
Added a new event wcpay_transactions_download_csv_in_browser. Registered
via PR #2718 in tracks-events-registration repo.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

shendy-a8c committed:
Merge branch 'develop' into update/9969-poll-async-job

shendy-a8c committed:
WIP get rid of browser disputes CSV download.

shendy-a8c committed:
Changelog.

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `10300` or branch name `update/7188-immediate-csv-download-for-disputes` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=update/7188-immediate-csv-download-for-disputes) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: b41a5b5417a4de753bdd647cfd0b57e5dcf0b70e
- Build time: 2025-02-12 09:38:53 UTC

_Note: the build is updated when a new commit is pushed to this PR._

shendy-a8c committed:
Rest end-point to get the disputes export URL.

jessy-p committed:
Merge branch 'develop' into update/9969-poll-async-job

Commit:
Address review comments

- Removed the file URL check that blocks local testing
- Changed isDownloading to isExportInProgress
- Removed the track event as it is not user triggered.

Commit:
Merge branch 'update/9969-poll-async-job' of https://github.com/Automattic/woocommerce-payments into update/9969-poll-async-job

shendy-a8c committed:
Merge branch 'update/9969-poll-async-job' into update/7188-immediate-csv-download-for-disputes

Commit:
Changed to a single notice.

shendy-a8c committed:
Merge branch 'update/9969-poll-async-job' into update/7188-immediate-csv-download-for-disputes

shendy-a8c committed:
Show notice and rename isDownload to isExportInProgress.

shendy-a8c committed:
Merge branch 'develop' into update/9969-poll-async-job

shendy-a8c committed:
Fix unit test after createNotice() call for success case is removed from useReportExport().

shendy-a8c committed:
Merge branch 'update/9969-poll-async-job' into update/7188-immediate-csv-download-for-disputes

shendy-a8c committed:
Fix unit test by removing tests that tries to look at CSV downloaded in the browser.

jessy-p commented on client/disputes/index.tsx:
```diff
@@ -530,6 +433,14 @@ export const DisputesList = (): JSX.Element => {
 		<Page>
 			<TestModeNotice currentPage="disputes" />
 			<DisputesFilters storeCurrencies={ storeCurrencies } />
+			{ isExportInProgress && (
```
New approach has been tried in #10211  based on design review. Please have a look, test and once that is approved and merged, we can use that here. 

The new approach is to remove this notice, use a single snackbar to inform that the export has started and would be downloaded automatically and emailed. There is no additional snackbar when the polling is completed.

shendy-a8c committed:
Merge branch 'develop' into update/7188-immediate-csv-download-for-disputes

# Conflicts:
#	changelog/update-9969-poll-async-job
#	client/disputes/test/index.tsx
#	client/hooks/use-report-export.ts
#	client/transactions/list/index.tsx
#	includes/admin/class-wc-rest-payments-transactions-controller.php
#	includes/wc-payment-api/class-wc-payments-api-client.php

shendy-a8c committed:
Update changelog to just a comment.

shendy-a8c committed:
Adapt to the latest UI and UX changes around CSV download experience from the transactions page.

shendy-a8c committed:
Re-apply the change to get disputes export URL API.

shendy-a8c committed:
Remove unused import.

shendy-a8c commented:
@jessy-p I have picked up the same notice changes from https://github.com/Automattic/woocommerce-payments/pull/10211.

I also made the changelog from this PR just a comment since we can just use one changelog entry for all the CSV experience change, no need one entry for each transactions, disputes, and payouts.

jessy-p committed:
Merge branch 'develop' into update/7188-immediate-csv-download-for-disputes

jessy-p reviewed (approved):
Tested with 
- successful export and download of file in browser ✅ 
- failed export and file did not download in browser, only emailed ✅ 
- Validated the implementation of notices ✅
